### DISCUSSION TITLE
Pass into shadow root

### DISCUSSION THREAD

**Discussion by Lomacar** [alpinejs/alpine#2205] [3. Feature ideas]
Is there a way to get Alpine to work on custom elements with a shadow root? Or is that a feature that could be implemented?

Maybe I just need to trigger the right things with the right timing? The way I currently have things, the custom elements are "upgraded" or "connected" or whatever you call it, and then Alpine is initiated. x-data and stuff works on the custom element but not inside the shadow root. However, after the page is loaded I can "see" Alpine inside the shadow root, for example, if I `console.log(Alpine)` from an onclick event. I could run something in the connectedCallback for the custom element, and I suppose I could even pass data from Alpine on the outside to the inside, but how do I make Alpine run inside the shadow root then?

It would be nice if Alpine simply had a built in way to force it to propagate into shadow roots.

**Comment by SimoTod** [comment]
There is a basic support in v3. 
Try to call Alpine.initTree(shadowRoot) on the connect callback. Because it's a custom element, the mutation observer won't be able to see what happens inside the shadowDom so i think you need to manually initialise it.

**Reply by Lomacar** [reply]
Oh great, thank you!

**Reply by Lomacar** [reply]
This doesn't exactly make the shadow root transparent. But I can manually move (x-)data inside before calling initTree.

Is there a performance concern if you are running Alpine every time inside a bunch of custom elements?

**Reply by SimoTod** [reply]
You shouldn't need it but you need to be careful about the timing. If your shadow dom needs to access a scope defined outside of the component, you need to make sure that the external component is ready. You can listen for the `alpine:initialized` event to make sure that everything else on the page is ready. In this way, you don't have to care about the order of your scripts.
For example, see https://codepen.io/SimoTod/pen/gOxOgGm

No particular concerns about calling initTree on the shadow Dom, it's what Alpine does anyway for standard components.

**Reply by Lomacar** [reply]
Thanks!

**Reply by shashank-brightness** [reply]
@SimoTod 
This does not work when using hmr api of bundler like webpack. I tried this when html changes and tried updating html in shadow dom followed by `Alpine.initTree(shadowRoot)` but i am getting error can not redefine $nextTick. Any idea on that?

I tried creating new shadow but even somehow can not remove shadowRoot to new shadow root. I need to refresh html in the shadow root on the fly.

**Reply by SimoTod** [reply]
@shashank-brightness hmr is a dev feature. I'm not particularly familiar with it but from their documentation they explain that it takes configuration or custom loaders to make it work.
I don't think it's compatible with Alpine out of the box.

**Reply by shashank-brightness** [reply]
@SimoTod HMR is already working but I was figuring out a way to reload html in the shadow dom once the dependency changes but it is not working properly. I need some react kind of workflow which can inject html changes automatically when someone updates template and so on...to be able to work with speed but unfortunately my most of the time is going to things like how to make it work properly.


kind of like this.

 ```
 const loadBrightnessContrastHTML = () => import(/* webpackMode: "eager" */'./brightnessContrast.html');
import initStyles from '../../initStyles';
import initHTML from '../../initHTML';
initHTML('brightness-contrast',loadBrightnessContrastHTML);

import { hotReloadAlpineComponent } from '../../hotReloader';
 
init() {
        initStyles(this.$el.shadowRoot);

        if(module.hot) {
            module.hot.accept('./brightnessContrast.html', function() {
                hotReloadAlpineComponent(this.$el.getRootNode().host ,loadBrightnessContrastHTML)
            }.bind(this));

           
        }
    },
```
and this is the hotReloader module

```
export function hotReloadAlpineComponent(componentEL, updateComponentHTML) {
    updateComponentHTML().then(componentHTML => {


        const alpineInitPointEl = componentEL.shadowRoot ? componentEL.shadowRoot : componentEL;

        

        console.log(componentHTML.default)
        const replacedComponent = new DOMParser().parseFromString(
            componentHTML.default,
            'text/html'
          ).body.firstChild;


        alpineInitPointEl.innerHTML = '';
        alpineInitPointEl.appendChild(replacedComponent);

        Alpine.initTree(alpineInitPointEl);

      
    });
}

```
I am using alpine js with component based setup where i import js object and html from template and then connect them together. I know Alpine never published documentation about its own `initTree` method in the doc but i exactly needed that and have seen somewhere so using it with shadow dom, but when html is again injected and initTree again called on shadowRoot it is giving error about $nextTick is redefining. Even if i don't call initTree on shadowRoot but on it's first child element(replacedComponent)  still same error pops up.
Doing some R&D now on virtual dom to make it work here does not seem to be achievable so i am switching back to older technology `livereload` now.

**Reply by ekwoka** [reply]
You can just swap the whole html and let it all cascade from there.

You can do `destroyTree` first before `initTree`.
